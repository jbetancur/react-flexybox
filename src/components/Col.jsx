import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { media, genWidth } from '../utils/media';

const AgnosticComponent = ({
  alignSelf,
  flex,
  xs,
  sm,
  md,
  lg,
  gutter,
  columnDivisions,
  order,
  minWidth,
  debug,
  component,
  ...rest
}) => {
  const Tag = component;

  return <Tag {...rest} />;
};

const ColStyle = styled(AgnosticComponent)`
  box-sizing: border-box;
  margin: ${props => props.gutter}px;
  max-height: 100%;
  order: ${props => props.order};
  align-self: ${props => props.alignSelf};
  ${props => props.debug && 'border: 1px solid green'};
  ${props => props.flex && media.flex`
    --width: ${genWidth(props.flex, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};

  /* if flex is just true then set flex to auto */
  ${props => !Number.isInteger(props.flex) && props.flex && 'flex: 1'};
  ${props => Number.isInteger(props.xs) && media.xs`
    --width: ${genWidth(props.xs, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.sm) && media.sm`
    --width: ${genWidth(props.sm, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.md) && media.md`
    --width: ${genWidth(props.md, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.lg) && media.lg`
    --width: ${genWidth(props.lg, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};

  /* Allows min-width to be overridden */
  ${props => props.minWidth && `min-width: ${props.minWidth}px`};
`;

const Col = ({
  alignSelf,
  flex,
  xs,
  sm,
  md,
  lg,
  gutter,
  columnDivisions,
  order,
  className,
  style,
  minWidth,
  debug,
  component,
  children,
  ...rest,
}) => (
  <ColStyle
    alignSelf={alignSelf}
    flex={flex}
    xs={xs}
    sm={sm}
    md={md}
    lg={lg}
    gutter={gutter}
    columnDivisions={columnDivisions}
    order={order}
    className={className}
    style={style}
    minWidth={minWidth}
    debug={debug}
    component={component}
    {...rest}
  >
    {children}
  </ColStyle>
);


Col.propTypes = {
  /**
   * react/html elements
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),

  /**
   * aligns the whole structure according to its value and has no effect when items are in a single line
   */
  alignSelf: PropTypes.oneOf(['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch']),

  /**
   * inherited from `Row`
   */
  gutter: PropTypes.number,

  /**
   * inherited from `Row`
   */
  columnDivisions: PropTypes.number,

  /**
   * The component to render the `Col` as
   */
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

  /**
   * inherited from `Row`
   */
  debug: PropTypes.bool,

  /**
   * allows overriding the `Col` item order
   */
  order: PropTypes.number,

  /**
   * sets the min-width on a `Col` item - this overrides `Row` `minColWidths` for a specific `Col`
   */
  minWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * overide the `className` on the root element of Row
   */
  className: PropTypes.string,

  /**
   * override the `style` on the root element of Row
   */
  style: PropTypes.object,

  /**
   * sets the default flex-basis and min-width percentage of the flex-item
   */
  flex: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 0-599-px
   */
  xs: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 600-959px
   */
  sm: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 960-1280px
   */
  md: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 1280px or greater
   */
  lg: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),
};

Col.defaultProps = {
  alignSelf: 'auto',
  children: [],
  order: 0,
  minWidth: false,
  className: '',
  style: {},
  component: 'div',
  flex: false,
  xs: false,
  sm: false,
  md: false,
  lg: false,
  gutter: 0,
  columnDivisions: 12,
  debug: false,
};

Col.displayName = 'Col';

export default Col;
