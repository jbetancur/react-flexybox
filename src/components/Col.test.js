import React from 'react';
import { render } from 'react-testing-library';
import 'jest-styled-components';
import Col from './Col';

describe('Col', () => {
  test('it mounts <Col />', () => {
    const { container } = render(<Col />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Col with flex: auto', () => {
    const { container } = render(<Col flex />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it overrides Col flex when xs', () => {
    const { container } = render(<Col flex={6} xs={12} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it sets size props', () => {
    const { container } = render(
      <Col
        xs={12}
        sm={6}
        md={3}
        lg={1}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it overrides default Col minWidth with minWidth prop', () => {
    const { container } = render(
      <Col
        flex={12}
        minWidth={42}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts <Col component="span" />', () => {
    const { container } = render(
      <Col component="span" />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it overrides the default component as a component', () => {
    const Faux = () => <div />;
    const { container } = render(
      <Col component={Faux} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it renders children', () => {
    const { container } = render(
      <Col><section>stuff</section></Col>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
