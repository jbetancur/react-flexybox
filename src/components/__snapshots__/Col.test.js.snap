// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col it mounts <Col /> 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

<div
  class="c0"
/>
`;

exports[`Col it mounts <Col component="span" /> 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

<span
  class="c0"
/>
`;

exports[`Col it mounts Col with flex: auto 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  --width: calc(8.333333333333332% - 0px);
  -webkit-flex: 0 1 var(--width);
  -ms-flex: 0 1 var(--width);
  flex: 0 1 var(--width);
  min-width: var(--width);
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
  class="c0"
/>
`;

exports[`Col it overrides Col flex when xs 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  --width: calc(50% - 0px);
  -webkit-flex: 0 1 var(--width);
  -ms-flex: 0 1 var(--width);
  flex: 0 1 var(--width);
  min-width: var(--width);
}

@media (min-width:0) and (max-width:599px) {
  .c0 {
    --width: calc(100% - 0px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

<div
  class="c0"
/>
`;

exports[`Col it overrides default Col minWidth with minWidth prop 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  --width: calc(100% - 0px);
  -webkit-flex: 0 1 var(--width);
  -ms-flex: 0 1 var(--width);
  flex: 0 1 var(--width);
  min-width: var(--width);
  min-width: 42px;
}

<div
  class="c0"
/>
`;

exports[`Col it overrides the default component as a component 1`] = `<div />`;

exports[`Col it renders children 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

<div
  class="c0"
>
  <section>
    stuff
  </section>
</div>
`;

exports[`Col it sets size props 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

@media (min-width:0) and (max-width:599px) {
  .c0 {
    --width: calc(100% - 0px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

@media (min-width:600px) and (max-width:959px) {
  .c0 {
    --width: calc(50% - 0px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

@media (min-width:960px) and (max-width:1280px) {
  .c0 {
    --width: calc(25% - 0px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

@media only screen and (min-width:1280px) {
  .c0 {
    --width: calc(8.333333333333332% - 0px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

<div
  class="c0"
/>
`;
